kiwmi/desktop/output.c

```diff
@@ -376,8 +388,132 @@ new_output_notify(struct wl_listener *listener, void *data)
+                configuration, output->wlr_output);
+
+        struct wlr_box *area = wlr_output_layout_get_box(
+            desktop->output_layout, output->wlr_output);
+
+        if (area != NULL) {
+            head->state.x = area->x;
+            head->state.y = area->y;
+        }
+    }
+
+    wlr_output_manager_v1_set_configuration(
+        desktop->output_manager, configuration);
+}
+
+static void
+output_manager_configure(
+    struct kiwmi_server *server,
+    struct wlr_output_configuration_v1 *configuration,
+    bool test)
+{
+    bool okay = true;
+    struct wlr_output_configuration_head_v1 *head;
+    wl_list_for_each (head, &configuration->heads, link) {
+        struct wlr_output *wlr_output = head->state.output;
+        struct kiwmi_output *output   = wlr_output->data;
+
+        output->enable_changed = head->state.enabled != wlr_output->enabled;
+
+        wlr_output_enable(wlr_output, head->state.enabled);
+        if (head->state.enabled) {
+            if (head->state.mode) {
+                wlr_output_set_mode(wlr_output, head->state.mode);
+            } else {
+                wlr_output_set_custom_mode(
+                    wlr_output,
+                    head->state.custom_mode.width,
+                    head->state.custom_mode.height,
+                    head->state.custom_mode.refresh);
+            }
+
+            wlr_output_layout_move(
+                server->desktop.output_layout,
+                wlr_output,
+                head->state.x,
+                head->state.y);
+            wlr_output_set_transform(wlr_output, head->state.transform);
+            wlr_output_set_scale(wlr_output, head->state.scale);
+        }
+
+        if (!wlr_output_test(wlr_output)) {
+            okay = false;
+            break;
+        }
+    }
+
+    wl_list_for_each (head, &configuration->heads, link) {
+        if (okay && !test) {
+            struct kiwmi_output *output = head->state.output->data;
+            if (head->state.enabled) {
+                if (~output->wlr_output->pending.committed
+                    & WLR_OUTPUT_STATE_ENABLED) {
+                    wl_signal_emit(&output->desktop->events.new_output, output);
+                } else {
+                    wl_signal_emit(&output->events.resize, output);
+                }
+            } else if (output->enable_changed) {
+                wl_signal_emit(&output->events.destroy, output);
+            }
+            wlr_output_commit(head->state.output);
+        } else {
+            wlr_output_rollback(head->state.output);
+            if (!okay) {
+                break;
+            }
+        }
+    }
+
+    if (okay) {
+        wlr_output_configuration_v1_send_succeeded(configuration);
+        wlr_output_manager_v1_set_configuration(
+            server->desktop.output_manager, configuration);
+    } else {
+        wlr_output_configuration_v1_send_failed(configuration);
+        wlr_output_configuration_v1_destroy(configuration);
+    }
```

Comment:
I’d split this condition into one checking `okay` and sending success/failure and another checking `okay && !test` and setting/destroying the config. This way it would get sent to clients even on successful test.